# Supplement: {#Supplement}

## Sequence Alignment, Normalization, and Transformation:

### Sequence Alignment

### Normalization and Transformation

### Heatmaps by condition and by strain

## The linear model of Condition and Stress identity accurately describes gene expression data
(or at least for most genes)

#### I hope to show that:

1. What's happening: I'm going to build the linear model
2. I'm going to show that it has good performance over most genes.

### Building the models:

For every gene, we build a linear model to predict gene expression.
The inputs to this model are the sample metadata (Condition and Kinase inhibition).
Although sometimes unrealistic, these models assume that changes in gene expression can be computed by adding up the individual effects provided by the condition, and from the kinase.
 For example:

For a gene that goes up 4X in from YPD to salt (+2 log2), and goes up 2X when PKA is inhibited (+1 log2), the model would expect the gene to be 8X increased in both Salt and PKA (+3 log2), or

$$\Delta E_g \approx C_g + K_g$$
Where $\Delta E$ is the change in log expression, and $C$ and $K$ are the contributions from the condition and kinase respectively.


```{r ref.label='create_linear_models', eval = FALSE}
```

In order to asses the quality for each fit, lets calculate the $R^2$, or the percentage of variance explained for each gene from the model.

```{r ref.label= 'model_quality', eval=FALSE}
```

Lets look at the $R^{2}$ for each gene in the full model:

```{r}
genes %>% select(Gene, performance) %>% unnest() -> model_quality

model_quality %>%
    ggplot(aes(x=r.squared)) +
    geom_density(fill='blue', alpha=.5) +
    ggtitle('Percentage of variance explained by the full model')
```

 Okay, so for most genes, the model does well.
For what genes does it perform badly?

Lets look at an example of strong and week performance:

```{r, echo=FALSE}
library(stringr)
library(gridExtra)
library(RGraphics)
library(ggrepel)

expression_plot = function(gene, measurements, t2g) {
orf = t2g$Gene[t2g$name == gene]
description = t2g$description[t2g$name == gene]
dat = measurements %>% dplyr::filter(Gene == orf)
p_heatmap = ggplot(dat, aes(y=Condition, x = Strain)) +
            geom_tile(aes(fill=Expression)) +
            theme(axis.text.x = element_text(angle = 90, hjust = 1))

p_cond_box = ggplot(dat, aes(x = Condition, y = Expression)) +
             geom_boxplot() +
             geom_text_repel(aes(label = condition_outlier), size=2, na.rm = TRUE) +
             coord_flip() +
             theme(axis.text.x = element_blank())


p_strain_box = ggplot(dat, aes(x = Strain, y = Expression)) +
               geom_boxplot() +
               geom_text_repel(aes(label = strain_outlier), size=2, na.rm = TRUE) +
               theme(axis.text.x = element_text(angle = 90, hjust = 1))

text = paste0(gene, " (", orf, "): ", description)
p_desc = ggplot() +
 annotate("text", x = 4, y = 25, size=8, label = text)

grid.arrange(p_heatmap + theme(legend.position="none", axis.text.x=element_blank(), axis.text.y=element_blank()),
          p_cond_box,
          p_strain_box + theme(legend.position="none", axis.text.y=element_blank()),
          textGrob(do.call(paste, c(as.list(strwrap(text, width = 0.7 * getOption("width"))), sep="\n"))), nrow=2, ncol=2)
}

residual_plot = function(gene, all, t2g) {
  orf = t2g$Gene[t2g$name == gene]
  dat = all %>% dplyr::filter(Gene == orf) %>% separate(sample_id, c('Condition', 'Strain'), sep = '_')

  p_heatmap1 = ggplot(dat %>% filter(residual_type == 'Base'), aes(x = Strain, y = Condition, fill = residual)) +
                geom_tile() +
                scale_fill_gradient2(low = ('cyan'), mid = 'black', high = 'yellow') +
                theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
                labs(fill = "Residual") +
                ggtitle("Expression ~ 1")

  p_heatmap2 = ggplot(dat %>% filter(residual_type == 'Strain'), aes(x = Strain, y = Condition, fill = residual)) +
                geom_tile() +
                scale_fill_gradient2(low = ('cyan'), mid = 'black', high = 'yellow') +
                theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
                labs(fill = "Residual") +
                ggtitle("Expression ~ Strain")

  p_heatmap3 = ggplot(dat %>% filter(residual_type == 'Condition'), aes(x = Strain, y = Condition, fill = residual)) +
                geom_tile() +
                scale_fill_gradient2(low = ('cyan'), mid = 'black', high = 'yellow') +
                theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
                labs(fill = "Residual") +
                ggtitle("Expression ~ Condition")

  p_heatmap4 = ggplot(dat %>% filter(residual_type == 'Full'), aes(x = Strain, y = Condition, fill = residual)) +
                geom_tile() +
                scale_fill_gradient2(low = ('cyan'), mid = 'black', high = 'yellow') +
                theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
                labs(fill = "Residual") +
                ggtitle("Expression ~ Strain + Condtion")



  grid.arrange(p_heatmap1, #+ theme(legend.position="none", axis.text.x=element_blank(), axis.text.y=element_blank()),
               p_heatmap2, #+ theme(legend.position="none", axis.text.x=element_blank(), axis.text.y=element_blank()),
               p_heatmap3, #+ theme(legend.position="none", axis.text.x=element_blank(), axis.text.y=element_blank()),
               p_heatmap4) #+ theme(legend.position="none", axis.text.x=element_blank(), axis.text.y=element_blank()))
}

```


```{r, warning=FALSE, message=FALSE, eval=FALSE}
model_quality %>% left_join(t2g) %>%
    select(Gene, name, r.squared) %>%
    arrange(r.squared) %>% head(5) %>% knitr::kable(caption = 'Worst Models')
```


```{r, message=FALSE, warning=FALSE}
model_quality %>% left_join(t2g) %>%
    select(Gene, name, r.squared) %>%
    arrange(-r.squared) %>% head(5) %>% knitr::kable(caption = 'Best Models')
```



Lastly, lets take a look at HSP12:

<!-- \blandscape -->

<!-- ```{r, fig.width=11, fig.height=8, echo=FALSE} -->
<!-- expression_plot('HSP12', measurements, t2g) -->
<!-- ``` -->


<!-- ```{r, fig.width=11, fig.height=8, echo=FALSE} -->
<!-- residual_plot('HSP12', std_residuals, t2g) -->
<!-- ``` -->

<!-- \elandscape -->



What makes the model weak? for very low scores (<.3) we generally see that this is dominated by genes with low expression values.

For moderately weak models, We would like to think it is because our simplistic linear model cannot account for epistatic effects between conditions and kinases.
But $R^2$ is a metric for the entire model.
We will later look at the residuals of individual data points to better detect epistasis.
For example FUS1 and HSP12 have very high $R^2$ even thought they have strong epistatic evidence.

Potential way to prove that this model is actually performing well:
What is the distribution of R^2 for row shuffled, column shuffled, or completely shuffled inputs?

\newpage





<!-- What does strains give you, that conditions dont? -->
<!-- break down ESR -->

\newpage
## Model recapitulates know gene expression modules
* Clustering of model coefficients reveal known gene expression modules
* Promoter Motifs
* Clusters in INSPIRE
* Identification of Kinase TF interactions.

This section is missing

\newpage


## Condition and Strain identities explain non-redundant aspects of the gene expression
Since a kinase inhibition is a much more targeted perturbation compared to a stress, we should ask how much information is in each contribution.

To break this down, lets see how much $R^2$ we loose by removing a variable from the model:

$$ R^2_{loss} = R^2_{full} - R^2_{reduced} $$

```{r ref.label='relative_importance', eval=FALSE}
```


```{r, eval=FALSE}
gather(as_tibble(lasts), Last_Factor, added_r2) %>%
  ggplot(aes(x = added_r2, fill = Last_Factor)) +
  geom_density(alpha = .5) +
  ggtitle('R^2 loss from removing variable', subtitle = 'Each variables contribution when included in the model last, also sometimes called usefulness')
```
## Standardized Residuals are normally distributed


Now that we have the standardized residuals, lets take a look at them.
What structure do they have? lets take a deep dive.

First question, what is the spread of these standardized residuals, according to theory, they should look similar to a z-score, meaning that they are normally distributed.
lets check:

```{r}
resids = std_residuals  %>% pull(residual)
hist(resids, breaks = 200)
```

Okay, that looks good, but what are the mean and variance? we expect 0, and 1.

```{r}
print(paste('Mean:', mean(resids)))
print(paste('Standard Deviation:', sd(resids)))

```

Close enough!


## Needs Placement


I don't know much about SLT2, and PCK1, other than they are in the same pathway.
Lets take a look at them.
(residual cutoff at 2)

```{r, warning=FALSE}
p = top_residuals %>%
  filter(abs(residual) > 2) %>%
  filter(Strain_Code %in% c('SLT2', 'PKC1')) %>%
  select(name, residual, Strain_Code, Condition) %>%
  spread(Strain_Code, residual) %>%
  mutate_at(3:4, function(x) if_else(is.na(x),0,x)) %>%
  ggplot(aes(x=SLT2,y=PKC1,fill=Condition, text= paste(name, Condition, sep='_'))) + geom_point()
# plotly::ggplotly(p, width = 880)
print(p)
```
Next, lets see what `Condition + Strain` pairs contain significant residuals over all genes in the ESR:

```{r}
top_residuals %>%
  filter(abs(residual) > 3) %>%
  filter(Gene %in% c(scan('../../input/genes_of_interest_and_gene_sets/ESR/repressed_ESR.txt', what = character()),
                     scan('../../input/genes_of_interest_and_gene_sets/ESR/activated_ESR.txt', what = character()))) %>%
  select(Condition, Strain_Code) %>%
  table() %>%
  `+`(1) %>%
  log() %>%
  heatmap(col=viridis::viridis(20))

```


#### Some quick observations here:
* Kinases in the same pathway cluster together:
    + SLT2 and PKC are very closely clustered!
    + PBS2 and HOG1 are very closely clustered!
* SCH9 Seems to have an effect, but only in YPD?

#### Diving into a some examples

Okay, so if the findings above are true, we should see that the residuals between PBS2, and HOG1 to correlate for genes that have significant residuals.
lets check that now:

```{r, warning=FALSE}
p = top_residuals %>%
    filter(abs(residual) > 3) %>%
    filter(Strain_Code %in% c('HOG1', 'PBS2')) %>%
  select(name, residual, Strain_Code, Condition) %>%
  spread(Strain_Code, residual) %>%
  mutate_at(3:4, function(x) if_else(is.na(x),0,x)) %>%
  ggplot(aes(x=HOG1,y=PBS2,fill=Condition,color = Condition, text= paste(name, Condition, sep='_'))) + geom_point()
print(p)
#plotly::ggplotly(p, width = 880)
```

Interestingly, inactivation of PBS2, and HOG1 do not always produce the correlated residuals.
They do for Salt, but do not for Menadione, and Glucose Dropout.
This might be evidence that PBS2 and HOG1 have different functions in non-salt conditions.
