<!-- Title -->
# An analysis of stress propagation through the kinome reveals ...

<!-- Claims / Results -->
## Results:
<!-- 6 point I want to make -->
<!-- What are the data required to make that point -->
<!-- What is the visualization necessary to convey that point. -->


### A strategy to identify Kinome wide signaling under stress
![TSNE](img/tsne_base.png)
![cell](img/Experiment_Cartoon.png)
![f1](img/Figure_1.png)


### Growth assays identify kinase contribution to fitness in a stress dependent manner  <!-- Push to suplement -->

### The full linear model of Condition and Stress identity accurately describes gene expression data

Lets look at the $R^{2}$ for each gene

![Full_model_performance](img/full_model_performance.png)


What is the R^2 for this data vs data that is row shuffled, column shuffled, or completely shuffled?

<!-- Still need to do this @TODO -->

### Condition and Strain identities explain non-redundant aspects of the gene expression

![Increased R2](img/added_r2_as_last.png)

### Model recapitulates know gene expression modules
* Clustering of model coefficients reveal known gene expression modules
* Promoter Motifs
* Clusters in INSPIRE

#### Identification of Kinase TF interactions.

### Discrepancies in the linear model predict epistatic relationships
![Salt Example](img/HOG_PBS_RESID.png)

#### residual profiles cluster into functionally enriched modules:
<!-- (so do the coefficients from before?) -->
*  GO
*  MEME/Scope

#### Error modules (groups of error profiles) reveal Known classes of epistatic relationships
<!-- Class 1 -->
The linear model assumes that Kinase identity and Condition identity make independent contributions to gene expression.  

1. If a Condition's effect propagates through a inhibition given Kinase, then the linear model will overestimate gene expression for that Condition-Kinase pair  
    * HSP12 on Salt + PKA (Get GD?)
2. If a Kinase's activity is required to correctly regulate a gene's expression in a condition specific manner, the linear model will badly estimate that condition specific interaction (as there is no term to correct for such a fine tuned response)
	  * Mating pathway  
	  
INSPIRE Cluster 36 (Contains FUS1 and other mating genes)

![Cluster 36](img/cluster36.png)

### Pbs2 inactivation inhibits activation of (a submodule of) the ESR

### Summary.
-----------------------------------------------------------------------

## Figure Layout:

### Figure 1:
#### Purpose:
The intention of the first figure is to do the following:
* Introduce the reader to the data
* Explain how Analog sensitive Kinases work
* Explain the scientific method (Stress + Drug, on strain, then RNAs-seq)
* Give the user a sense for the vast size of the dataset

#### Panels:
1. A cartoon of stress induced kinase signaling leading to activation of a transcriptional program. Also the use of an ATP analog that specifically blocks the activity of a kinase.
2. A heatmap or other method showing all the data
3. Principal Components of the data?



### Figure 2:
Validation of known biology
* Can we show that kinases regulate known target genes
* Can we show that kinases regulate known transcription factors and transcriptional programs
* Can we show that stresses activate known target genes
* can we show that stresses activate known transcription factors and transcriptional programs



### Figure 3 and 4:
* Kinase centric analysis
* Maybe Kinase to TF hypothesized connectivity graph?
* What genes are target of a given kinase?
	* what conditions are required for this to occur?
* Can we back calculate the TF from the gene?

### Figure 4:

### Figure 5:
Potential new hypotheses

Compile a list of new biology for others to potentially go after
1. Hog1 activity represses Hug1 - a reporter for DNA damage
2. AZC up-regulates the levels of Act1 (Actin)
3. Pbs2 activity is requried to repress the Ribosome biogenesis pathways under stress.

### Figure 6:
Validation of a vignette

## Figure Methods:

### Single gene analysis
* Heatmap/Barplot Plots

### Single Kinase analysis
* Volcano Plots
* Scope and MEME

### Single Stress analysis
* Morpheous and others

## Notes
One thing that I've been thinking about, is that I will be using the same methods to show known biology, as well as show new biology. therefore, isn't it a better idea to go with a methods based approach and follow a generic structure of showing the known biology and then hypothesizing on new biology each time?
