## A strategy to identify Kinome wide signaling under stress

### Experimental Pipeline:

![](img/Experiment_Cartoon.png)

### Analysis Pipeline:

#### Load Data:

Data is aligned with STAR, and then expression levels are normailzed using DESeq's variance stabalizing transformation. It is then loaded in for analysis.

```{r, ref.label=data_loading, eval=FALSE}
```

#### Calculation of statistics, and filtering:

We then calculate basic statistics on the gene expression data:

```{r ref.label=gene_statistics, eval=FALSE}
```

Genes with no variability will be problematic in downstream analysis. As we can infer very little about these genes, it is apprpriate to filter them out. We will arbitrarily filter out genes whos dynamic range is less than 1 in all the samples

```{r ref.label=gene_filter, eval=FALSE}
```


### Full dataset heatmap

#### Ordered by condition:

![](img/by_condition.png)

#### Ordered by kinase:

![](img/by_strain.png)


### Clustering of samples shows distinct kinases, and condition groups.

There are 300 samples, each with an applied condition, kinase inhibition, and a gene expression vector. Here, we cluster the samples by performming t-SNE on gene expression vectors with default t-SNE parameters


```{r ref.label='tsne_cluster', eval=FALSE}
```


When plotting the gene expression vectors in t-SNE space, we see that the samples cluster into distinct kinase, and condition clusters:

```{r test, message = FALSE}
library(ggrepel)

ggplot(sample_tsne,
       aes(x=TSNE1, 
           y=TSNE2, 
           color = Condition, 
           label = Strain)) +
  geom_point(size=4) +
  geom_text_repel(size = 2,
                  color = 'black',
                  point.padding = NA,
                  box.padding = unit(0.01, "lines"))

```


heirarchically clustering the t-SNE projections:

```{r, echo = -1, message=FALSE}
library(dendextend)

dend = sample_tsne %>% 
  select(starts_with('TSNE')) %>% # Select the TSNE variables 
  dist() %>% # Euclidean distance
  hclust() %>% # Heirarchical Clustering
  as.dendrogram()

the_bars = cbind(factor(sample_tsne$Condition), 
                 factor(sample_tsne$Strain))

plot(dend)
colored_bars(colors = the_bars, 
             dend = dend, 
             sort_by_labels_order = TRUE)

```



![](img/Kinase_Pipeline.png)

