---
title: "Salt"
output: html_notebook
---

Here we perform an individual analysis of the Salt Condition.

# Extract the data:

```{r}
library(tidyverse)
select = dplyr::select
library(magrittr)
library(edgeR)
#library(DEFormats)
meta = read_csv('../../intermediate/meta_clean.csv') %>%
  mutate(Strain = relevel(factor(Strain), ref = 'WT'),
         Condition = relevel(factor(Condition), ref = 'YPD'))
t2g = read_csv('../../intermediate/t2g.csv')

meta_wt = meta %>% filter(Strain == 'WT')
counts_wt = readDGE(meta_wt$star_path, columns = c(1,3), labels = meta_wt$Experiment)[-c(1:3),]
meta_tm = meta %>% filter(Condition == 'Salt')
counts_tm = readDGE(meta_tm$star_path, columns = c(1,3), labels = meta_tm$Experiment)[-c(1:3),]

```


# Defining the Salt Program in WT cells 
```{r}
library(DESeq2)
dds_wt = DESeqDataSetFromMatrix(counts_wt, colData = meta_wt, ~Condition)
dds_wt %<>% DESeq()

res_wt = results(dds_wt, 
                 name = "Condition_Salt_vs_YPD",
                 independentFiltering = FALSE)

res_wt = lfcShrink(dds_wt,
                   coef = "Condition_Salt_vs_YPD",
                   res = res_wt)

plotMA(res_wt)

```

Okay, so there are some differentially expressed genes. what are they?

## Repressed in Salt Genes:
```{r}
fdr.threshold = 0.1



get_go_terms_from_list = function(query_list, background_list) {
    library(goseq)
    library(org.Sc.sgd.db)
    genes = numeric(length(background_list))
    names(genes) = background_list
    genes[query_list] = 1
    pwf = nullp(genes,"sacCer2","ensGene")
    GO.wall=goseq(pwf,gene2cat=as.list(org.Sc.sgdGO2ALLORFS))
    return(GO.wall)
}

res_wt %>% 
  as.data.frame %>% 
  rownames_to_column('Gene') %>% 
  filter(padj<fdr.threshold) %>% 
  left_join(t2g) %>% 
  select(log2FoldChange, padj, Gene, name) %>% 
  filter(log2FoldChange<0) %>% 
  pull(Gene) %T>% head() %>% 
  get_go_terms_from_list(rownames(res_wt)) %>% 
  as_tibble() %>%
  filter(over_represented_pvalue < 0.05 | under_represented_pvalue < 0.05) %>%
  arrange(desc(numDEInCat/numInCat)) %>% select(term, starts_with('num'), ends_with('pvalue'))
```


## Activated Salt Genes:
```{r}

res_wt %>% 
  as.data.frame %>% 
  rownames_to_column('Gene') %>% 
  filter(padj<fdr.threshold) %>% 
  left_join(t2g) %>% 
  select(log2FoldChange, padj, Gene, name) %>% 
  filter(log2FoldChange>0) %>% 
  pull(Gene) %T>% head() %>% 
  get_go_terms_from_list(rownames(res_wt)) %>% 
  as_tibble() %>%
  filter(over_represented_pvalue < 0.05 | under_represented_pvalue < 0.05) %>%
  arrange(desc(numDEInCat/numInCat)) %>% select(term, starts_with('num'), ends_with('pvalue'))
```

# Extract gene expression changes between WT and Kinases:
```{r}
dds_tm = DESeqDataSetFromMatrix(counts_tm, colData = meta_tm, ~Strain)
dds_tm %<>% DESeq()

res_tm = resultsNames(dds_tm)[-1] %>%
  map_dfr(~ results(dds_tm,
                    name = .x,
                    independentFiltering = FALSE) %>%
            {lfcShrink(dds_tm, coef=.x, res = ., parallel=TRUE)} %>% # Use when betaPrior is FALSE
            as.data.frame %>%
            as_tibble(rownames = 'Gene') %>%
            mutate(term = .x)) %>%
  mutate(Strain = word(term,2,sep = '_'))


  res_wt %>% as.data.frame() %>%
  as_tibble(rownames='Gene') %>% 
  select(log2FoldChange, padj, Gene) %>% 
  dplyr::rename(wt_l2FC = log2FoldChange, wt_padj = padj) %>% 
  mutate(movement = sign(wt_l2FC)) %>%
  right_join(res_tm) -> joined_res
```


## Probing the relationship between kinase changes and condition changes:

Lets plot the Kinase changes, vs the Salt change:

```{r}
joined_res %>% filter(wt_padj < 0.1) %>% ggplot(aes(x = wt_l2FC, y = log2FoldChange, group = movement)) + geom_point() + geom_smooth(method='lm') + facet_wrap(~Strain) + xlab('Wild Type Fold Change (Salt vs. YPD)') + ylab('Strain Differece (Strain vs Wild Type, all in Salt)')

```
It looks like in Salt, inactivation of most kinases have very little effect.

Some kinases have an effect:
IRE1
TPK123
YPK1
PBS2
KSP1

However these changes do not seem to corrolate with the WT response. Except for IRE1

IRE1 seems to be negatively corrolated with the WT Salt vs. YPD response. 

I interpret this to mean that the IRE1 Strain is in not only modifying gene expression when inactivated (as is the case for the above kianses), but its effect is directly inhibiting the cells ability to mount its WT response to Salt. One thing I think is very supprising is that IRE1 is not known to target the whole genome, but rather the UPR gene module. However we see this global trened. 

Lets pull out the slopes:

```{r}
library(broom)
fits = joined_res %>% filter(wt_padj<0.3) %>%
  group_by(Strain, movement) %>% nest() %>% 
  mutate(fit = map(data, ~lm(log2FoldChange~wt_l2FC - 1, data = .x)),
         coef = map(fit, tidy),
         glance = map(fit, glance),
         data_aug = map(fit, augment),
         slope = map_dbl(coef, ~ filter(., term == 'wt_l2FC') %>% pull(estimate)),
         slope_conf = map_dbl(coef, ~ filter(., term == 'wt_l2FC') %>% pull(p.value)),
         r_squared = map_dbl(glance, 'adj.r.squared')
         )
         
ggplot(fits, aes(y=slope,x=r_squared,label=Strain, color = movement)) + geom_label()        
```


## Are there patterns in the kinase changes? do some kinases do the same thing in Salt?

Lets use a heatmap to detect genes that change in the same way between kinases:

```{r}
as_matrix <- function(x, rowname_col = 1){
   if(!tibble::is_tibble(x) ) stop("x must be a tibble")
  y <- as.matrix.data.frame(x[,-rowname_col])
  rownames(y) <- x[[rowname_col]]
  y
}

library(ComplexHeatmap)
library(dendsort)

exp_matrix = res_tm %>% 
  select(Gene, Strain, log2FoldChange) %>%
  spread(key = Strain, value = log2FoldChange) %>% na.omit() %>%
  as_matrix

exp_col = res_wt %>% as.data.frame() %>% as_tibble(rownames = 'Gene') %>%
    select(Gene, log2FoldChange) %>%
  as_matrix

exp_col = exp_col[rownames(exp_matrix),]

col_dend = dendsort(hclust(dist(t((exp_matrix)[]))))
p = draw(
  Heatmap(exp_matrix,
          show_row_names = F,
          name = 'Kinase vs. WT',
          cluster_columns = col_dend,
          column_dend_reorder = FALSE,
          km=5
          #cluster_rows = FALSE,
          #row_order = order(exp_col), 
          #row_dend_reorder = F
          ) +
  Heatmap(exp_col,
          show_row_names = F,
          name = paste0("Tunicamyin vs. YPD")#,
          #col = colorRamp2(c(min(res[sub_genes,]$log2FoldChange),
          #                   0,
          #                   max(res[sub_genes,]$log2FoldChange)), c("blue", "#EEEEEE", "red"))
          )
)

```

lets now make the same plot, but subset the genes that we believe are differentially expressed in Tm compared to YPD.

Scatter plots with filtering
```{r}
joined_res %>% filter(wt_padj<.1) %>% ggplot(aes(x = wt_l2FC, y = log2FoldChange, group = sign(wt_l2FC))) + geom_point() + geom_smooth(method='lm') + facet_wrap(~Strain) + xlab('Wild Type Fold Change (Salt vs. YPD)') + ylab('Strain Differece (Strain vs Wild Type, all in Salt)')
```


```{r}
exp_matrix = res_tm %>% 
  select(Gene, Strain, log2FoldChange) %>%
  spread(key = Strain, value = log2FoldChange) %>% na.omit() %>%
  as_matrix

exp_col = res_wt %>% as.data.frame() %>% as_tibble(rownames = 'Gene') %>%
    select(Gene, log2FoldChange) %>%
  as_matrix

de.genes = res_wt %>% 
  as.data.frame() %>% 
  as_tibble(rownames = 'Gene') %>% 
  filter(padj<0.1 & abs(log2FoldChange)>.5) %>% pull(Gene)

exp_col = exp_col[de.genes,]
exp_matrix = exp_matrix[de.genes,]

  

col_dend = dendsort(hclust(dist(t((exp_matrix)[]))))
p = draw(
  Heatmap(exp_matrix,
          show_row_names = F,
          name = 'Kinase vs. WT',
          cluster_columns = col_dend,
          column_dend_reorder = FALSE,
          km=6
          #cluster_rows = FALSE,
          #row_order = order(exp_col), 
          #row_dend_reorder = F
          ) +
  Heatmap(exp_col,
          show_row_names = F,
          name = paste0("Tunicamyin vs. YPD")#,
          #col = colorRamp2(c(min(res[sub_genes,]$log2FoldChange),
          #                   0,
          #                   max(res[sub_genes,]$log2FoldChange)), c("blue", "#EEEEEE", "red"))
          )
)
```

It seems like inactivating many Kinases have an effect, meaning that those kinases must be active in Salt. However, not all of those kinases have an effect that is corrolated with The Salt response. It does seem like IRE1 is anticorrolated.s

```{r}
# all in log space, so subtraction = division in count space
strain_difference = IRE1_Tm - WT_Tm # What does inactivating IRE1 do in Tm compared to WT 
condition_difference = WT_Tm - WT_YPD # what does adding Salt do to WT cells

plot(x_axis = condition_difference,
     y_axis = strain_difference)
```

